### ISSUE TITLE
globals – make their performance less awful

### ISSUE THREAD

**Issue by StefanKarpinski** [JuliaLang/julia#8870]
This keeps coming up, and rather than direct people to the performance tips, I feel like we should really just try to fix the issue. One relatively non-disruptive approach might be to emit code that accesses globals with a fast path that assumes that they have the same type that they currently have and a slow path that handles arbitrary type. A harder fix would be the compile-time version of that: generate code that assumes that globals retain their current type and invalidate it if the type changes. One might want to let compilation happen a couple of times before giving up and generating pessimistic general code.

Related: #265, #524, #964

**Comment by Sisyphuss** [comment]
Suppose I write:
```
param = 1
myfunc(param)
```

Since now I have a global variable, will I get a relatively poor performance?

Or should I write instead:

```
function main()
  param = 1
  myfunc(param)
end

main()
```

**Comment by carnaval** [comment]
The first version is fine. It is the lookup from inside compiled code which is slow, since the compiler cannot know a priori the type of the result.
If you call `func(param)` the function will be compiled for an `Int` parameter, and no global lookup will happen inside.

**Comment by Sisyphuss** [comment]
Here's my experiment, and I find no significant performance difference in version 0.3.7
```
param = 0.001
@time sleep(param)
```
```
function main()
  param = 0.001
  @time sleep(param)
end
main()
```
Both take 2.1 ms.

**Comment by JeffBezanson** [comment]
Think of it this way: *accessing* a global variable is slow. So the existence of a global variable in your code might not matter if you only read it once. You don't want to read or write a global variable repeatedly in a loop. But if you read its value once and pass it to a function, there's no problem.

In this case the overhead of `sleep` will be much bigger than anything else; you won't get meaningful times.

**Comment by Sisyphuss** [comment]
The remark of @JeffBezanson  seems reasonable. I did this experiment: 
```
@time begin
  p = 1
  for i in 1:100000
    p += 1
  end
end
```
elapsed time: 0.002861372 seconds (1591840 bytes allocated)
```
function main()
  @time begin
    p = 1
    for i in 1:100000
      p += 1
    end
  end
end

main()
```
elapsed time: 1.676e-6 seconds (0 bytes allocated)

But what remains mysterious to me is that why the compiler can't treat global variables just in the same as local variables. And how to explain the huge difference of memory allocation in the above example?

-------------------
By the way, an anonymous function version is below:
```
f = () -> begin
  p = 1
  for i in 1:100000
    p += 1
  end
end

@time f()
```
elapsed time: 3.422e-6 seconds (80 bytes allocated)

**Comment by JeffBezanson** [comment]
Globals are different because they can be observed anywhere when they change. To convert a global to a local, the compiler would have to prove that no possible code anywhere in the system can see the changing value of `p` during the loop.

**Comment by timholy** [comment]
The memory allocation is from boxing. Even the type of the global might change; if it did, and you didn't box things, you'd get a segfault.

**Comment by Sisyphuss** [comment]
Actually, I don't know how the compiler treats local or global variables. I'm not from the computer science community. But I try my best to guess what you mean. It seems to me that you are worrying about the _multi-threading_ situation.

**Comment by timholy** [comment]
No, just this:
```jl
julia> globvar = 0
0

julia> function foo()
           s = 0
           for i = 1:5
               bar()
               s += globvar
           end
       end
foo (generic function with 1 method)

julia> function bar()
           global globvar
           globvar = "oops"
           nothing
       end
bar (generic function with 1 method)

julia> foo()
ERROR: MethodError: `+` has no method matching +(::Int64, ::ASCIIString)
Closest candidates are:
  +(::Any, ::Any, ::Any)
  +(::Any, ::Any, ::Any, ::Any...)
  +(::Int64, ::Int64)
  ...
 in foo at none:5
```

**Comment by Sisyphuss** [comment]
@timholy, typically your example only illustrates how dangerous global variables could be and it is not a good programming habit. But I try to understand what you implied. You imply that if the compiler had treated the `globvar` in the same way as a local variable, this run-time error wouldn't have been detected. Isn't it?

**Comment by Sisyphuss** [comment]
If this is the case, I further deduce that the compiler interpret a local variable as a **fix** address in the memory, and a global variable as a name in a "variable list".

**Comment by timholy** [comment]
Exactly. Julia creates optimized code that depends upon how the machine represents information: the code for adding two `Float64`s is different from the code for adding two `Float32`s or two `Int`s. If you're wrong about the representation, you either get complete junk or a segfault, depending on the nature of the code.

"Boxing" just means wrapping the machine representation with some additional information that allows you to ask, "what type of object is this, anyway?" It takes memory, however.

**Comment by pao** [comment]
Memory for the box, time to handle the indirection (and to ask the "what are you" question), and possibly time to perform a dynamic dispatch, depending on context. It's turtles all the way down when types are unpredictable.

**Comment by Sisyphuss** [comment]
Thanks for your reply!

I have just found an interesting phenomenon: 
```
@time begin
  p = 1
  for i in 1:510
    p += 1
  end
end
```
elapsed time: 2.9961e-5 seconds (0 bytes allocated)
```
@time begin
  p = 1
  for i in 1:511
    p += 1
  end
end
```
elapsed time: 2.8285e-5 seconds (16 bytes allocated)

Within 510 iterations, there are no extra memory cost. From the 511th iteration, every extra iteration takes 16 bytes more. Does it mean that there is no boxing within 510 iterations?

**Comment by StefanKarpinski** [comment]
The first 512 integers (0-511) are preallocated and boxed versions of them are always kept around. After that you need to actually allocate a new object.

**Comment by Sisyphuss** [comment]
Now I understand. Let's go back to our original discussion about why to avoid global variables. 

I think what you are really against is not the global variable itself, but **loop involving global variables**, or in other words, **using global variables in any loop**.

**Comment by StefanKarpinski** [comment]
If you use a global variable in some function and then that function gets called in a loop, you have the same problem – and if you're a library author, there's no way to predict how your function will get called. So I don't see how the "in a loop" qualification really helps. The only way you can be sure that your code won't be called in a loop is if it happens once at global scope, in which case, sure it's fine to use globals.

**Comment by Sisyphuss** [comment]
@StefanKarpinski, it is generally considered as a bad habit to use global variables in any function. It is so for *all* languages, not specifically for Julia. Therefore, if I write the functions myself, or the library is written by some skilled developer, I can safely use global variables by following my previous remark. Moreover, my remark helps _quantify_ the overhead of the global variable, which was considered as _demon_ previously.

**Comment by Sisyphuss** [comment]
In order to better _quantify_ the overhead, I permit myself to ask this question:
If I write all Julia codes in the global scope, is the performance going to be as poor as writing in other languages (matlab, R, python...) ? Of course, without calling C/Fortran functions.

**Comment by simonster** [comment]
If you use a variable whose type depends on a global variable in a loop, you will still have a performance problem, even if you don't use the global variable itself in a loop. You can see this quite clearly with `@code_typed`/`@code_warntype`. There is no mystery, just type inference.

If you write your code in the global scope, performance will most likely be worse than Python, which has a highly optimized interpreter, and less overhead for calling functions than Julia when types are not known (because Python has no generic functions). I don't know how it would compare to MATLAB and R, which have substantially worse interpreters.

**Comment by Sisyphuss** [comment]
Thanks @simonster , your example is persuasive. It makes me have to change my remark to **avoiding loop involving unknown type (variables)** .

About your second remark, em... writing Julia code has a risk!

**Comment by timholy** [comment]
@Sisyphuss, the thing you should realize: this is basically a trap only for newcomers. If you decide to stick with julia, I predict that you'll look back on this and wonder what you were so worried about.

A good step is to get to know the tools we have for diagnosing performance problems. See http://docs.julialang.org/en/release-0.3/manual/performance-tips/. Of course, we're always interested in people who make improvements to those tools. I know @tonyhffong is really hoping to find some collaborators to work on his excellent [Lint.jl](https://github.com/tonyhffong/Lint.jl), which can warn you about this very problem.

**Comment by Sisyphuss** [comment]
Thanks @Tim, I have read the manual once, and I'm reading it for the second time. I don't _decide_ to stick with Julia; I _already_ stick with her. I'm not worried about it; I just try to understand her.

The package you mentioned is fine for learning, but I am already using Julia in my current research work.

**Comment by pao** [comment]
@Sisyphuss you can't abbreviate GitHub usernames--this may notify other users on GitHub that are not a part of this discussion and possibly annoy them. Thanks!

**Comment by jiahao** [comment]
I just ran into a rather curious version of this issue while implementing a character streaming file parser.

Here are 3 naive ways of iterating through a file character by character:

```jl

f = open(filename)

@time while !eof(f)
  read(f, UInt8)
end

seek(f, 0)

@time while true
   read(f, UInt8)
   eof(f) && break
end

seek(f, 0)

@time while true
    try
        read(f, UInt8)
    catch e
        isa(e, EOFError) ? break : rethrow(e)
    end
end

close(f)
```

For a particular 200MB file, I get these timings

| version | timing |
| --- | --- |
| `while !eof` | 36.055 seconds |
| `eof && break` | 18.663 seconds |
| `catch EOFError` | 16.732 seconds |

If I wrap the exact same code in a function and call it, I get

| version | timing |
| --- | --- |
| `while !eof` | 3.740 seconds |
| `eof && break` | 3.526 seconds |
| `catch EOFError` | 6.927 seconds |

This example demonstrates that timing comparsions in global scope can have absolutely no correlation between the same comparisons done in a function, which arguably is much worse than just having uniformly slow globals.

**Comment by jakebolewski** [comment]
Is this surprising?  Globals are in part slow because they inhibit optimization within the compiler.  Predicting which optimizations are applied by the compiler is always perilous, so this is not really a gotcha.

**Comment by jiahao** [comment]
> Is this surprising? 

Perhaps not to those of us jaded by trying to second-guess the compiler, but I'd think that most people would find this surprising.

**Comment by jakebolewski** [comment]
This is not second guessing, it is just a fact of life with a JIT compiler.

**Comment by jiahao** [comment]
> this is not really a gotcha.

I can't see how this attitude is justifiable. New users unaccustomed to the subtleties of the Julia compiler are the ones most likely to write such code in global scope. And this is not a new observation.

**Comment by StefanKarpinski** [comment]
This makes the "first encounter experience" with the language bad for many people, and there are things we can do to fix it.

**Comment by ssfrr** [comment]
This is probably a dumb idea, but would it be possible to instrument access to globals and then print out a warning if Julia sees a global accessed a bunch of times in a short time period? That way we'd still guide people in the right direction if they're likely hitting a performance bottleneck, but not annoy folks playing round in the REPL in global scope.

**Comment by StefanKarpinski** [comment]
It's possible, but I still feel it would be better to simply fix the problem. People do want to do computations in the REPL and have it not be horribly slow. It's also not that uncommon to write scripts that do stuff at global scope and while it's not a big deal, having to wrap it all in a function is annoying. I think that after all the ambiguity warnings, this is the lowest hanging fruit for making the "Julia experience" nicer.

**Comment by malmaud** [comment]
Would it be crazy to just not allow the type of globals to change?

**Comment by malmaud** [comment]
Or maybe have them type-const by default, with a keyword (`nonconst`?, `volatile`) to indicate that the type can change.

**Comment by StefanKarpinski** [comment]
Not at all: https://github.com/JuliaLang/julia/issues/964#issuecomment-7468492. But I suspect that it would be nicer to allow type annotations for now (partial mitigation of the issue) and then implement the optimistic type-stability assumption with invalidation of code that relies on the assumed type. It's more work, but it's a better experience for users.

**Comment by ScottPJones** [comment]
:+1: :100: to @StefanKarpinski 's cunning plan with the invalidation of code. Could the type annotations be done in time for 0.4? (maybe wishful thinking)

**Comment by ChrisRackauckas** [comment]
Is there any way to be able to use optimizations with literals in the global scope?

```julia
function f(x)
  x^2
end

@code_warntype f(2)
Variables:
  #self#::#f
  x::Int64

Body:
  begin 
      return (Base.mul_int)(x::Int64, x::Int64)::Int64
  end::Int64

@code_warntype 2^2
Variables:
  #self#::Base.#^
  x::Int64
  p::Int64

Body:
  begin 
      return $(Expr(:invoke, MethodInstance for power_by_squaring(::Int64, ::Int64), :(Base.power_by_squaring), :(x), :(p)))
  end::Int64
```

I know that `^` can handle integer literals and for example inline that `^2` is `x*x`, but it only does that in a function. In the global scope, wouldn't you still have enough information to do these optimizations, and wouldn't that at least help global scope loops a little bit?

Type assertions are also a really good idea, or at least a `typeconst`. It might seem like over-typing, but people are essentially using `const` to do the same thing already, so it might as well work correctly. `const` has other side-effects, making this essentially a trap for users don't realize `const` is not just a performance enhancer for constant type.

**Comment by StefanKarpinski** [comment]
Potential proposal (needs some work, but it's the most promising thing we could come up with on today's triage call):

1. Implement type annotations for globals (#964).
2. Globals without explicit type annotations are implicitly type-const.
3. Error if an explicit global type annotation is violated from anywhere.
4. Error if an implicit global type annotation is violated from non-top-level code.
5. Warn if an implicit global type annotation is violated from top-level code.

For 1.0 only the behavioral part of this has to be implemented, but this scheme allows making globals fast with some relatively straightforward optimizations. Most globals don't change type, and this allows efficient code to be generated under that assumption. If those type assumptions are violated from the top-level, all methods relying on those assumptions can be invalidated. This allows relatively convenient use at the REPL, while still fixing the global performance problem.

**Comment by ChrisRackauckas** [comment]
>Error if an implicit global type annotation is violated in non-top-level code.

So this is saying that if a function tries to modify a global in a non-type-stable way, it will error. That sounds good. But

>Warn if an implicit global type annotation is violated in top-level code.

So then

```julia
x = 2
x = 2.0
```

that will warn every time?

**Comment by StefanKarpinski** [comment]
`x = 2; x = 2.0` would warn, yes. I'm not sure what you mean by "every time" – it would warn every time you evaluate it at the top-level, yes.

**Comment by vtjnash** [comment]
After discussion on triage, I propose we only implement item 1 (type annotations for globals, #964), and close this issue immediately. There doesn't seem to be any consensus that the other proposals here are necessary, or even uniformly useful.

**Comment by StefanKarpinski** [comment]
The triage winds have turned here and there are serious usability and feasibility concerns about this. Everyone agrees that #964 should be implemented and with that you can conveniently express both that a global is constant and that it is type-constant. Performance could be fixed without changing the current semantics with clever on-stack-replacement since you could just invalidate and replace any method whose code depends on the type that is changed.

**Comment by StefanKarpinski** [comment]
There's still a performance issue here, we just decided not to make any semantic changes.

**Comment by ViralBShah** [comment]
With typed globals in 1.8, I am hoping it is ok to close this.

https://julialang.org/blog/2022/08/julia-1.8-highlights/#typed_globals

**Comment by antoine-levitt** [comment]
Can you keep it open? While I understand it's not a priority, it would be awesome if something like

> Performance could be fixed without changing the current semantics with clever on-stack-replacement

suggested by @StefanKarpinski could eventually be implemented to make quick&dirty script writing performant.

**Comment by StefanKarpinski** [comment]
Could we union-split in the current type of globals at the time of method generation? I.e. every global access implicitly looks like this:
```jl
if typeof(G) === typeG
    # fast path
else
    # slow path
end
```
That doesn't involve any code invalidation. It trades slowness for a bit of generated code bloat.

**Comment by vtjnash** [comment]
We could, but it is probably beyond pointless to the level of actually slowing down the code, since there typically is no actual cost in the dispatch here, but there is usually cost in using the return value due to the boxing and dispatch implied in not being able to ascertain the type of that PhiNode.

**Comment by StefanKarpinski** [comment]
So is there anything we could realistically do to make globals faster at this point? We have typed globals now, which solves much of the original issue (either make your globals `const` or typed). The only thing I can think of at this point would be assuming type-constness of all globals initially and invalidating and code generated on that assumption if a global is assigned a different type.

**Comment by Keno** [comment]
My current thinking on this is the following:

In toplevel scope, assignment to globals without a declared type automatically widens the type restriction to encompass both the current restriction and the newly assigned value. Assignments in local scope retain the current semantics. I think this gives us basically everything we want.

In order to implement this we will need to finish the bpart logic for globals, which was punted from 1.12, and since it's a breaking semantic change, we'll want to implement the opt-in mechanism for language evolution.
